src/server/worker/datum_iterator.go

```diff
@@ -0,0 +1,285 @@
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+)
+
+// DatumIterator is an interface which allows you to iterate through the datums
+// for a job. A datum iterator keeps track of which datum it is on, which can be Reset()
+// The intended use is by using this pattern `for di.Next() { ... datum := di.Datum() ... }`
+// Note that since you start the loop by a call to Next(), the datum iterator's location starts at -1
+type DatumIterator interface {
+	Reset()
+	Len() int
+	Next() bool
+	Datum() []*Input
+}
+
+type pfsDatumIterator struct {
+	inputs   []*Input
+	location int
+}
+
+func newPFSDatumIterator(pachClient *client.APIClient, input *pps.PFSInput) (DatumIterator, error) {
+	result := &pfsDatumIterator{}
+	if input.Commit == "" {
+		// this can happen if a pipeline with multiple inputs has been triggered
+		// before all commits have inputs
+		return result, nil
+	}
+	fs, err := pachClient.GlobFileStream(pachClient.Ctx(), &pfs.GlobFileRequest{
+		Commit:  client.NewCommit(input.Repo, input.Commit),
+		Pattern: input.Glob,
+	})
+	if err != nil {
+		return nil, err
+	}
+	for {
+		fileInfo, err := fs.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		result.inputs = append(result.inputs, &Input{
+			FileInfo:   fileInfo,
+			Name:       input.Name,
+			Lazy:       input.Lazy,
+			Branch:     input.Branch,
+			EmptyFiles: input.EmptyFiles,
+		})
+	}
+	// We sort the inputs so that the order is deterministic. Note that it's
+	// not possible for 2 inputs to have the same path so this is guaranteed to
+	// produce a deterministic order.
+	sort.Slice(result.inputs, func(i, j int) bool {
+		// We sort by descending size first because it can boost performance to
+		// process the biggest datums first.
+		if result.inputs[i].FileInfo.SizeBytes != result.inputs[j].FileInfo.SizeBytes {
+			return result.inputs[i].FileInfo.SizeBytes > result.inputs[j].FileInfo.SizeBytes
+		}
+		return result.inputs[i].FileInfo.File.Path < result.inputs[j].FileInfo.File.Path
+	})
+	// make sure it gets initialized properly
+	result.Reset()
+	return result, nil
+}
+
+func (d *pfsDatumIterator) Reset() {
+	d.location = -1
+}
+
+func (d *pfsDatumIterator) Len() int {
+	return len(d.inputs)
+}
+
+func (d *pfsDatumIterator) Datum() []*Input {
+	return []*Input{d.inputs[d.location]}
+}
+
+func (d *pfsDatumIterator) Next() bool {
+	d.location++
+	return d.location < len(d.inputs)
+}
+
+type unionDatumIterator struct {
+	iterators []DatumIterator
+	unionIdx  int
+}
+
+func newUnionDatumIterator(pachClient *client.APIClient, union []*pps.Input) (DatumIterator, error) {
+	result := &unionDatumIterator{}
+	for _, input := range union {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		result.iterators = append(result.iterators, datumIterator)
+	}
+	result.Reset()
```

Comment:
Same as above, but I think this would have to be deferred. Not really a big deal, but it might be be a good idea to future proof.